### PR TITLE
Allow users to update their configs and quick launch sessions

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
There is a new page under `/configs` which allows the user to update their configuration. This will poll for the current list of verified `desktop` types and display them in a list. There is some special handling when the default desktop has not been verified.

In addition, a `Quick launch session` button/link has been added to the nav bar. Currently it will re-create a session each time the user connects. Possible this should be extended to re-connect to an existing session.

### CODE REVIEW

File: src/ConfigsPage.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+      </h4>
+      <div className="card-body container">
+        <Form>
+          <FormGroup row>
+            <Label for="desktop" sm={2} size="lg">Desktop</Label>
+            <Col>
+              <Input  type="select" name="desktop" id="desktop" required
+                      value={desktop} onChange={e => updater(setDesktop, e)}>
+                <DisabledDesktopOption desktops={desktops} desktop={configs.desktop}/>
+                <DesktopOptions desktops={desktops}/>
```
Comment (benarmston) [inline src/ConfigsPage.js:76]: This drop down doesn't seem quite right.  It's not clear that the disabled option is the default.  If I change the desktop type but then want to change back it isn't entirely clear how I select that default again.

There are a couple of options to address this:

1. Remove the disabled option and use the `selected` attribute on `option`.  Add a simple indicator to the default desktop, perhaps `*` or `(default)`.
2. Switch over to [React Select](https://react-select.com/home). Remove the disabled option; use the `defaultValue` prop; add nice styling to the default option.

Can you take a look at one of the above options please?  Option (1) is perfectly fine if (2) is awkward FSR.

File: src/ConfigsPage.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+              <Input  type="number" name="geometry-x" id="geometry-x" required
+                      placeholder="X Geometry" value={xGeometry} onChange={e => updater(setXGeometry, e)}
+              />
+            </Col>
+            <span sm="auto" size="lg" className="align-middle">X</span>
+            <Col>
+              <Input  type="number" name="geometry-y" id="geometry-y" required
+                      placeholder="Y Geometry" value={yGeometry} onChange={e => updater(setYGeometry, e)}
+              />
+            </Col>
```
Comment (benarmston) [inline src/ConfigsPage.js:92]: We spoke about this briefly before.  The geometry for the desktop controls both the VNC session size and the X session size via `xrandr`.  The latter is done by [installing](https://github.com/openflighthpc/flight-desktop-types/blob/master/gnome/session.sh#L116) the [geometry script](https://github.com/openflighthpc/flight-desktop-types/blob/master/gnome/flight-desktop_geometry.sh).  Other desktops have something similar.

As these are intended for xrandr input, we should ensure that they are sensible values for xrandr.  As a first pass, hard coding  the following values would sufficient.  Later we're likely to want to determine which of these values are supported by the machine running, and to also given them better names, e.g., Tablet, Large Laptop, Mobile etc..


 * 1920x1200
 * 1920x1080
 * 1600x1200
 * 1680x1050
 * 1400x1050
 * 1360x768 
 * 1280x1024
 * 1280x960 
 * 1280x800 
 * 1280x720 
 * 1024x768 
 * 800x600  
 * 640x480  


Obviously, this change implies that a single select would be used rather than two inputs.

File: src/api.js [hunk 3]
```diff
@@ -75,6 +97,23 @@ export function useTerminateSession(id) {
...
+      headers: {
+        'Content-Type': 'application/json',
+        'Accepts': 'application/json',
+      },
+      cachePolicy: 'no-cache',
+    });
+  const patch = function(desktop, geometry) {
+    return request.patch({ desktop: desktop, geometry: geometry });
+  }
+  return { request, patch: patch };
```
Comment (benarmston) [inline src/api.js:114]: FYI: `{ variableName }` is equivalent to `{ variableName: variableName }`. So the above could be shortened to `{ request, patch }`.

**Review by benarmston — CHANGES_REQUESTED** [review]
Thanks for your work on this @WilliamMcCumstie .  Having a quick launch button is very nice.

I've added a few comments below for you to look at mostly related to styling and user-visible copy.

File: src/ConfigsPage.js [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+  return <Button color="success" className="pull-right" disabled={!modified} onClick={submit}>
+    Update Configuration
+  </Button>
+}
+
+function updateSuccessfulToast() {
+  let body = (
+    <div>
+      Your configurations have been updated!
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: 23a39a9a34a36b40279c64b864ac049188316388

File: src/api.js (hunk 3 above)
Comment (WilliamMcCumstie) [reply src/api.js]: ee2ce3f39ef2bee10cf438d003d1b97268b9eb6f

File: src/ConfigsPage.js [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    } else {
+      return <Layout configs={config_req.data} desktops={desktop_req.data.data} />
+    }
+  }
+}
+
+function Layout({ configs, desktops }) {
+  // Determine the current settings
+  const [x, y] = configs.geometry.split("x");
+  const d = desktops.map(d => { return d.id }).includes(configs.desktop) ? configs.desktop : null;
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: ee2ce3f39ef2bee10cf438d003d1b97268b9eb6f

File: src/ConfigsPage.js (hunk 1 above)
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: bddfb5814c6fe51681c01bba0002eaf310b153da

File: src/ConfigsPage.js [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+  // Returns the disabled element
+  return(<option key={desktop} disabled>{desktop}</option>);
+}
+
+function DesktopOptions({desktops}) {
+  return desktops.map(desktop => {
+    var element = null;
+    if (desktop.verified) {
+      element = <option key={desktop.id}>{desktop.id}</option>
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: f0104521d7eac6e188f363b6c74e33bbafe72380

File: src/ConfigsPage.js [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+        // The API response does not match the expected values
+        addToast(updateFailedToast({ errorCode: "did-not-update" }));
+      }
+    } else {
+      console.log("Failed to update configuration");
+      addToast(updateFailedToast({ errorCode: utils.errorCode(data) }));
+    }
+  }
+
+  return <Button color="success" className="pull-right" disabled={!modified} onClick={submit}>
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: cec78525aa443877e93726d6aa7a208f14ac14d4

File: src/ConfigsPage.js [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+          <FormGroup row>
+            <Label for="desktop" sm={2} size="lg">Desktop</Label>
+            <Col>
+              <Input  type="select" name="desktop" id="desktop" required
+                      value={desktop} onChange={e => updater(setDesktop, e)}>
+                <DisabledDesktopOption desktops={desktops} desktop={configs.desktop}/>
+                <DesktopOptions desktops={desktops}/>
+              </Input>
+            </Col>
+          </FormGroup>
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: 778c333eceb159bfb8657be94a54430609219f43

[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: 778c333eceb159bfb8657be94a54430609219f43

File: src/ConfigsPage.js (hunk 2 above)
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: c4e583830254f321b3aa553f9a52a79a079a1f3d

File: src/ConfigsPage.js [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+function DesktopOptions({desktops, selected, original}) {
+  return desktops.map(desktop => {
+    var element = null;
+    const is_selected = (selected === desktop.id);
+    const pretty  = prettyDesktopName[desktop.id];
+    const label = original === desktop.id ? `${pretty} (default)` : pretty
+    if (desktop.verified) {
+      element = <option value={desktop.id} label={label} selected={is_selected}/>
```
[the remark this answers is not in the document]
Comment (benarmston) [reply src/ConfigsPage.js:101]: Same for `GeometryOptions`.

File: src/ConfigsPage.js [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+  );
+}
+
+function DesktopOptions({desktops, selected, original}) {
+  return desktops.map(desktop => {
+    var element = null;
+    const is_selected = (selected === desktop.id);
+    const pretty  = prettyDesktopName[desktop.id];
+    const label = original === desktop.id ? `${pretty} (default)` : pretty
+    if (desktop.verified) {
```
Comment (benarmston) [inline src/ConfigsPage.js:100]: We should include all desktop types not just verified ones.  Its possible that a desktop type might be verified on the compute nodes but not on the head node.  We want to allow the user to set the default suitable for the compute nodes.

File: src/ConfigsPage.js [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    if (utils.errorCode(req.data) === 'Unauthorized') {
+      return <UnauthorizedError />;
+    } else {
+      return <DefaultErrorMessage />;
+    }
+  } else {
+    if (config_req.loading || desktop_req.loading) {
+      return <Loading />;
+    } else {
+      return <Layout configs={config_req.data} desktops={desktop_req.data.data} />
```
Comment (benarmston) [inline src/ConfigsPage.js:33]: The desktops should be extracted from the request via `flight-webapp-components` `utils.getResourcesFromResponse(data)` function.  It has some useful error handling in it.

File: src/ConfigsPage.js [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        </Form>
+      </div>
+    </div>
+  );
+}
+
+function DesktopOptions({desktops, selected, original}) {
+  return desktops.map(desktop => {
+    var element = null;
+    const is_selected = (selected === desktop.id);
```
Comment (benarmston) [inline src/ConfigsPage.js:97]: We don't need `is_selected` here.  The `value` property on the `select` is sufficient and the correct way of doing this.  I think I may have suggested using `selected` in a previous comment.

Also JS variables should be `camelCase` not `under_score`.

File: src/ConfigsPage.js [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+    }
+    return element;
+  });
+}
+
+function GeometryOptions({geometries, selected, original}) {
+  return geometries.map(geometry => {
+    const is_selected = (selected === geometry);
+    const label = original === geometry ? `${original} (default)` : geometry
+    return <option value={geometry} label={label} selected={is_selected}/>
```
Comment (benarmston) [inline src/ConfigsPage.js:111]: Again, we don't need `selected`.  We do need `key`.

File: src/ConfigsPage.js [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+  });
+}
+
+function UpdateButton({desktop, geometry, originalStruct, setOriginalStruct}) {
+  const { addToast } = useToast();
+  const { request, patch } = useUpdateUserConfig();
```
Comment (benarmston) [inline src/ConfigsPage.js:117]: FYI: this could be `const { request, response, patch } = ...` which would change a later line from `if (request.response.ok)` to `if (response.ok)`.  I think the latter reads a little nicer.

File: src/ConfigsPage.js [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+      } else {
+        // The API response does not match the expected values
+        addToast(updateFailedToast({ errorCode: "did-not-update" }));
+      }
+    } else {
+      console.log("Failed to update configuration");
+      addToast(updateFailedToast({ errorCode: utils.errorCode(data) }));
+    }
+  }
+  const is_original = (desktop === originalStruct.desktop) && (geometry === originalStruct.geometry);
```
Comment (benarmston) [inline src/ConfigsPage.js:138]: Can we have `isOriginal` instead of `is_original`.

File: src/ConfigsPage.js [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+
+  return <Button color="success" className="pull-right" disabled={request.loading || is_original} onClick={submit}>
+    { request.loading ? <i className="fa fa-spinner fa-spin mr-1"/> : null }
+    <span>{request.loading ? "Updating Configuration..." : "Update Configuration"}</span>
```
Comment (benarmston) [inline src/ConfigsPage.js:142]: This should be `Updating...` rather than `Updating Configuration...`.  That has better consistency with our other "action buttons".

File: src/ConfigsPage.js [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+      Unfortunately there has been a problem updating your configuration.
+      Please try again and, if problems persist, help us to more quickly
+      rectify the problem by contacting us and letting us know.
+    </div>
+  );
+
+  return {
+    body,
+    icon: 'danger',
+    header: 'Failed to update configuration',
```
Comment (benarmston) [inline src/ConfigsPage.js:172]: This should be "your configuration".  We refer to "your configuration" most everywhere else.

**Review by benarmston — CHANGES_REQUESTED** [review]
This is looking much better @WilliamMcCumstie .  I have a few more changes for you to look at please.

File: src/ConfigsPage.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: c5e32db319fcee07cfce4d8bbeb4c806bfcdd8f0

File: src/ConfigsPage.js (hunk 10 above)
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js:100]: This is an easy enough change, however IMO it is premature.

Currently the `quick launch` button uses this setting to fire up new desktops. Setting the default to an unverified desktop will break this feature.

This leaves one of three options:

1. Do nothing, `quick launch` will continue to operate correctly if configured via the webapp,
2. Conditionally disable `quick launch` (will likely require an additional request), or
3. Accept that `quick launch` may fail depending on the configuration.

---

## Option 3: Breaking Quick Launch and Poor UX

_Technically_ the `quick launch` button can fail if the user re-configures the `flight-desktop` via CLI. This in general, leads to a poor UX and should be avoided. It is however an "advance use case" and pre-existing issue.

However, re-configuring the `flight-desktop` via the webapp is a relatively simple use case. This should not lead to `quick launch` button breaking. This rules out option 3.

---

## Option 2: Disable Quick Launch

This option is do-able, but means each page load will need to determine if the default desktop has been verified.

This raises a few headaches as it will either require:

* An additional request to `flight-desktop-restapi` on every page load, or
* Some how cache the request and periodically update it.

Both of these options raise their own issues and ideally would be avoided.

tl;dr Option 2 is doable, but is a bit of a pain.

---

## Option 1: Do not allow unverified desktops

This is the do-nothing option, and accept the default can not be set to an unverified desktop. This has the following implications to the `quick launch` button:

* It should function correctly when configured via the `webapp`, and
* It may break when configured via the `CLI`.

Currently the `webapp` does not support launching desktops on remote machine, so the functionality should focus on the host. For this reason, the do nothing option is probably best.

At some point, we will likely need to address the question:
_What if remote nodes have different verified desktops?_

However this should be done in such a way not to break the `quick launch`. However these issues can be kicked down the road.

Comment (benarmston) [reply src/ConfigsPage.js:100]: You raise a good point about the "quick launch" button being coupled to this setting.  For now, add `disabled` `option`s for the desktop types that are reported as being unverified.  That's a better UX than them simply being omitted.

I'll have a think about how this should work for desktop types that have different verification states on the head node and compute nodes.

File: src/ConfigsPage.js (hunk 12 above)
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: 887b565482bc47b1c82d0bdc80b54c5c0655fd3f

File: src/ConfigsPage.js (hunk 13 above)
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: Done with `DesktopOptions` commits

File: src/ConfigsPage.js (hunk 14 above)
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: b2fd156eb9673a045392aee00d2160dcdcf72aea

File: src/ConfigsPage.js (hunk 17 above)
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: 35d894c5154787122650e1cb707a772cb6f48ea5

File: src/ConfigsPage.js (hunk 16 above)
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: 35d894c5154787122650e1cb707a772cb6f48ea5

File: src/ConfigsPage.js (hunk 15 above)
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: 35d894c5154787122650e1cb707a772cb6f48ea5

File: src/ConfigsPage.js (hunk 11 above)
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js:33]: The `desktop_req.data.data` exists because it is generally bad practice to return an `array` from an AP (Although I do believe the particular CSRF this pertains to has been resolved).

However the other GET requests do not have a `data` key. This has been a long standing implementation of `flight-desktop-restapi`.

---

As far as the additional error handling, the `response` is already checked to be `OK`. This should be sufficient to infer that `desktop_req.data.data` will be an array.

File: src/ConfigsPage.js (hunk 14 above)
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js:120]: This change does not work FSR. Assumable the `loading` variable would need to be waited on

I do not the resolve to dig through `useFetch` to work out why it doesn't work. So I have reverted the change to `request.response`:

dc3157fdb743163b3cc8f4b35438f3c48ba595fc

File: src/ConfigsPage.js (hunk 10 above)
Comment (WilliamMcCumstie) [reply src/ConfigsPage.js]: 2cc3a7edc0954febc37ad474fb3cb9e3643f4a43

File: src/ConfigsPage.js (hunk 14 above)
Comment (benarmston) [reply src/ConfigsPage.js:138]: Fair enough.

**Review by benarmston — APPROVED** [review]
You misunderstood what I was after with the `selected` attribute on `option`.  Instead of

```
<select value="something">
  <option value="something" selected="true">Something</option>
  <option value="something-else">Something else</option>
</selected>
```

we should instead have

```
<select value="something">
  <option value="something">Something</option>
  <option value="something-else">Something else</option>
</selected>
```

The `value` attribute on `select` determines which `option` is selected.  The `selected` attribute on `option` is ignored and therefore confusing.

I've made that change a few other stylistic changes.

**Comment by benarmston** [comment]
LGTM.
